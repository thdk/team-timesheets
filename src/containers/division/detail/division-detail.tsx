import React, { useCallback } from "react";
import { TextField } from "@rmwc/textfield";

import { Form, FormField } from "../../../components/layout/form";
import { FlexGroup } from "../../../components/layout/flex";
import IconPicker from "../../../components/icon-picker";

type Props = {
    readonly onNameChanged: (name: string) => void;
    readonly onIconChanged: (icon: string) => void;
    readonly icon?: string;
    readonly name?: string;
}

export const DivisionDetail = (props: Props) => {
    const {
        icon,
        name,
        onIconChanged,
        onNameChanged,
    } = props;

    const handleNameChanged = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
        onNameChanged(e.target.value);
    }, [onNameChanged]);

    const handleIconChanged = useCallback((icon: string) => {
        onIconChanged(icon);
    }, [onIconChanged]);

    return <>
        <Form>
            <FlexGroup extraCssClass="row">
                <FormField>
                    <TextField
                        outlined
                        label="Name"
                        value={name}
                        onChange={handleNameChanged}
                    />
                </FormField>
                <FormField first={false}>
                    <IconPicker
                        onChange={handleIconChanged}
                        value={icon}
                        outlined
                    />
                </FormField>
            </FlexGroup>
        </Form>
    </>;
};
DivisionDetail.displayName = "DivisionDetail";