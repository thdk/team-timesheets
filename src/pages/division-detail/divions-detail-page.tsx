import * as React from "react";
import copy from "copy-text-to-clipboard";
import { v4 as uuidv4 } from 'uuid';
import cryptoRandomString from 'crypto-random-string';

import { queue } from "../../components/snackbar";
import { withAuthentication } from "../../containers/users/with-authentication";
import { RedirectToLogin } from "../../routes/login";
import { DivisionDetail } from "../../containers/division/detail";
import { useEffect } from "react";
import { useViewStore } from "../../contexts/view-context";
import { useDivisionStore } from "../../contexts/division-context";
import { useUserStore } from "../../contexts/user-context";
import { goToUserProfile } from "../../routes/users/profile";
import { useRouterStore } from "../../stores/router-store";
import { transaction } from "mobx";
import { setChildNavigation, setTitleForRoute } from "../../internal";
import { IViewAction } from "../../stores/view-store";
import { observer } from "mobx-react-lite";

export const DivisionDetailPage = withAuthentication(
    observer(
        () => {
            const divisionStore = useDivisionStore();
            const viewStore = useViewStore();
            const userStore = useUserStore();
            const router = useRouterStore();

            useEffect(
                () => {
                    transaction(() => {
                        if (router.currentRoute) {
                            setChildNavigation({
                                view: viewStore,
                                navigateBack: () => goToUserProfile(router, "divisions"),
                            });
                            setTitleForRoute({
                                view: viewStore
                            },
                                router.currentRoute,
                            );
                        }

                        const actions: IViewAction[] = [
                            {
                                icon: { label: "Save", content: "save" },
                                shortKey: { key: "s", ctrlKey: true },
                                action: async () => {
                                    const {
                                        name,
                                        icon,
                                    } = divisionStore.activeDocument || {};

                                    // validation
                                    if (!name) { return; }

                                    if (divisionStore.activeDocument?.id) {
                                        divisionStore.updateDocument(
                                            {
                                                ...divisionStore.activeDocument,
                                            },
                                            divisionStore.activeDocument?.id,
                                        );
                                    }
                                    else {
                                        const divisionId = uuidv4();
                                        const divisionUserId = uuidv4();
                                        await Promise.all([
                                            divisionStore.collection.addAsync({
                                                name,
                                                icon,
                                                createdBy: userStore.authenticatedUserId!,
                                                id: divisionId,
                                            }, divisionId),
                                            divisionStore.divisionCodesCollection.addAsync({
                                                code: cryptoRandomString({ length: 6, type: 'distinguishable' }),
                                                divisionId,
                                            }),
                                            userStore.divisionUsersCollection.addAsync(
                                                {
                                                    ...userStore.authenticatedUser!,
                                                    divisionId,
                                                    roles: {
                                                        admin: true
                                                    },
                                                },
                                                divisionUserId,
                                            ),
                                        ]);

                                        if (!userStore.authenticatedUser?.divisionUserId) {
                                            userStore.updateAuthenticatedUser({
                                                divisionUserId,
                                                divisionId,
                                            });
                                        }
                                    }

                                    goToUserProfile(router, "divisions");
                                }
                            },
                        ];

                        if (divisionStore.activeDocument?.id) {
                            const divisionId = divisionStore.activeDocument?.id;
                            actions.push(
                                {
                                    icon: {
                                        content: "share",
                                        label: "Get division entry code",
                                    },
                                    action: () => {
                                        divisionStore.divisionCodesCollection.query = ref => ref.where("divisionId", "==", divisionId);
                                        divisionStore.divisionCodesCollection.fetchAsync()
                                            .then(() => {
                                                divisionStore.divisionCodesCollection.docs.forEach(d => {
                                                    queue.notify({
                                                        title: `Division entry code: ${d.data!.code}`,
                                                        dismissesOnAction: true,
                                                        actions: [
                                                            {
                                                                label: 'Copy',
                                                                icon: 'content_copy',
                                                                onClick: () => {
                                                                    copy(d.data!.code);
                                                                },
                                                            },
                                                        ],
                                                    });
                                                })
                                            });
                                    }
                                },
                            )
                        }

                        viewStore.title = router.currentRoute?.title || "";
                        viewStore.setActions(actions);
                    });


                    return () => {
                        viewStore.setActions([]);
                    };
                },
                [
                    viewStore,
                    divisionStore,
                    userStore,
                    divisionStore.activeDocument,
                ],
            );

            useEffect(
                () => {
                    if (router.params?.id === undefined) {
                        divisionStore.createNewDocument();
                    } else {
                        divisionStore.setActiveDocumentId(router.params?.id.toString());
                    }

                    return () => {
                        divisionStore.setActiveDocumentId(undefined);
                    }
                },
                [
                    divisionStore,
                    router,
                ]
            );
            return (
                <DivisionDetail />
            );
        }),
    <RedirectToLogin />,
);
