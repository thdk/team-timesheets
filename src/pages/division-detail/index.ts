export * from "./divions-detail-page";
