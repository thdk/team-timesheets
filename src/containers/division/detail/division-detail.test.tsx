import React from "react";
import { render, act, fireEvent } from "@testing-library/react";
import { DivisionDetail } from "./division-detail";

describe("DivisionDetail", () => {
    it("should render without data", () => {
        const {
            asFragment,
        } = render(
            <DivisionDetail
                onIconChanged={jest.fn()}
                onNameChanged={jest.fn()}
            />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should call onNameChanged", () => {
        const onNameChanged = jest.fn();
        const {
            container,
        } = render(
            <DivisionDetail
                onIconChanged={jest.fn()}
                onNameChanged={onNameChanged}
            />
        );

        act(() => {
            fireEvent.change(container.querySelector("input")!, {
                target: {
                    value: "Foo",
                },
            });
        });

        expect(onNameChanged).toHaveBeenCalledWith("Foo");
    });
})