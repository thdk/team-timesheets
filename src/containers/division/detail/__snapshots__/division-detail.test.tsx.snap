// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DivisionDetail should render without data 1`] = `
<DocumentFragment>
  <div
    class="form"
  >
    <div
      class="flex row horizontal"
    >
      <div
        class="first formField"
      >
        <label
          aria-labelledby="textfield-Name-label"
          class="mdc-text-field mdc-text-field--upgraded mdc-text-field--outlined"
        >
          <span
            class="mdc-text-field__ripple"
          />
          <input
            class="mdc-text-field__input"
            value=""
          />
          <div
            class="mdc-notched-outline mdc-notched-outline--upgraded"
          >
            <div
              class="mdc-notched-outline__leading"
            />
            <div
              class="mdc-notched-outline__notch"
            >
              <span
                class="mdc-floating-label"
                id="textfield-Name-label"
                style="transition-duration: 0s;"
              >
                Name
              </span>
            </div>
            <div
              class="mdc-notched-outline__trailing"
            />
          </div>
        </label>
      </div>
      <div
        class="formField"
      >
        <label
          aria-labelledby="textfield-Icon-label"
          class="mdc-text-field mdc-text-field--upgraded mdc-text-field--outlined mdc-text-field--with-trailing-icon"
        >
          <span
            class="mdc-text-field__ripple"
          />
          <input
            class="mdc-text-field__input"
            value=""
          />
          <div
            class="mdc-notched-outline mdc-notched-outline--upgraded"
          >
            <div
              class="mdc-notched-outline__leading"
            />
            <div
              class="mdc-notched-outline__notch"
            >
              <span
                class="mdc-floating-label"
                id="textfield-Icon-label"
                style="transition-duration: 0s;"
              >
                Icon
              </span>
            </div>
            <div
              class="mdc-notched-outline__trailing"
            />
          </div>
          <i
            class="rmwc-icon rmwc-icon--ligature material-icons mdc-text-field__icon mdc-text-field__icon--trailing"
            tabindex="0"
          >
            edit
          </i>
        </label>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
